R/susie_utils.R

```diff
@@ -1164,3 +1166,53 @@ run_refine <- function(data, best_model, rerun, tol_improve = 0) {
   }
   best_model
 }
+
+# Helper function to update variance components and derived quantities
+update_model_variance <- function(data, model, lowerbound, upperbound) {
+  variance_result <- update_variance_components(data, model)
+  model$sigma2 <- max(lowerbound, variance_result$sigma2)
+  model$sigma2 <- min(model$sigma2, upperbound)
+  
+  # Update additional variance components if they exist
+  if (!is.null(variance_result$tau2)) {
+    model$tau2 <- variance_result$tau2
+  }
+  
+  # Update derived quantities after variance component changes
+  data <- update_derived_quantities(data, model)
+  
+  # Transfer theta from data to model if computed (for non-sparse methods)
+  if (!is.null(data$theta)) {
+    model$theta <- data$theta
+    
+    # Update fitted values to include theta: XtXr = XtX %*% (b + theta)
+    b <- colSums(model$alpha * model$mu)
+    model$XtXr <- data$XtX %*% (b + model$theta)
+  }
+  
+  return(list(data = data, model = model))
+}
+
+# Create standard matrix initialization (shared by all backends)
+create_matrix_initialization <- function(p, L, scaled_prior_variance, var_y, residual_variance, 
+                                         prior_weights, include_non_sparse = FALSE) {
+  mat_init <- list(
```

Comment:
is this kind of the return object? if so, should we give it more informative name to imply it's part of the output?